src/sentry/releasehealth/metrics.py

```diff
@@ -150,3 +158,185 @@ def _compute_crash_free_rate(data: Dict[str, float]) -> Optional[float]:
+
+        def _count_sessions(total: bool, referrer: str) -> Dict[Any, int]:
+            query = Query(
+                dataset=Dataset.Metrics.value,
+                match=Entity("metrics_counters"),
+                select=[Column("value")],
+                where=_get_common_where(total)
+                + [
+                    Condition(Column("metric_id"), Op.EQ, metric_id(org_id, "session")),
+                ],
+                groupby=_get_common_groupby(total),
+            )
+
+            return _convert_results(
+                raw_snql_query(
+                    query,
+                    referrer=referrer,
+                    use_cache=False,
+                )["data"],
+                total=total,
+            )
+
+        def _count_users(total: bool, referrer: str) -> Dict[Any, int]:
+            query = Query(
+                dataset=Dataset.Metrics.value,
+                match=Entity("metrics_sets"),
+                select=[Column("value")],
+                where=_get_common_where(total)
+                + [
+                    Condition(Column("metric_id"), Op.EQ, metric_id(org_id, "user")),
+                ],
+                groupby=_get_common_groupby(total),
+            )
+
+            return _convert_results(
+                raw_snql_query(
+                    query,
+                    referrer=referrer,
+                    use_cache=False,
+                )["data"],
+                total=total,
+            )
+
+        # XXX(markus): Four queries are quite horrible for this... the old code
+        # sufficed with two. From what I understand, ClickHouse would have to
+        # gain a function uniqCombined64MergeIf, i.e. a conditional variant of
+        # what we already use.
+        #
+        # Alternatively we may want to use a threadpool here to send the
+        # queries in parallel.
+
+        # NOTE: referrers are spelled out as single static string literal so
+        # S&S folks can search for it more easily. No string formatting
+        # business please!
+
+        # Count of sessions/users for given list of environments and timerange, per-project
+        sessions_per_project: Dict[int, int] = _count_sessions(
+            total=True, referrer="releasehealth.metrics.get_release_adoption.total_sessions"
+        )
+        users_per_project: Dict[int, int] = _count_users(
+            total=True, referrer="releasehealth.metrics.get_release_adoption.total_users"
+        )
+
+        # Count of sessions/users for given list of environments and timerange AND GIVEN RELEASES, per-project
+        sessions_per_release: Dict[Tuple[int, int], int] = _count_sessions(
+            total=False, referrer="releasehealth.metrics.get_release_adoption.releases_sessions"
+        )
+        users_per_release: Dict[Tuple[int, int], int] = _count_users(
+            total=False, referrer="releasehealth.metrics.get_release_adoption.releases_users"
+        )
+
```

Comment:
Are we planning to cache this somehow? This to me seems like quite an expensive function call

Reply:
We don't know what data inconsistency caching would introduce. Let's wait until we have performance numbers before we start optimizing